File: 55-jump-game/jump-game.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canJump = function(nums) {
    if (nums.length <= 1) {
        return true;
    }

    let maximum = nums[0];
    let i = 0;

    while (i < nums.length) {
        if (maximum <= i && nums[i] === 0) {
            return false;
        }

        if (i + nums[i] > maximum) {
            maximum = i + nums[i];
        }

        if (maximum >= nums.length - 1) {
            return true;
        }

        i++;
    }

    return false;

};

File: 55-jump-game/jump-game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        max_reach = 0
        
        for i in range(len(nums)):
            if i > max_reach:
                return False  # Cannot reach index i
            max_reach = max(max_reach, i + nums[i])
            
            if max_reach >= len(nums) - 1:
                return True  # Can reach the end of the array
        
        return False
